google/cloud/bigtable/data/_cross_sync/_decorators.py

```diff
@@ -0,0 +1,441 @@
+        """
+        Helper to convert ast primitives to python primitives. Used when unwrapping arguments
+        """
+        import ast
+
+        if ast_node is None:
+            return None
+        if isinstance(ast_node, ast.Constant):
+            return ast_node.value
+        if isinstance(ast_node, ast.List):
+            return [cls._convert_ast_to_py(node) for node in ast_node.elts]
+        if isinstance(ast_node, ast.Tuple):
+            return tuple(cls._convert_ast_to_py(node) for node in ast_node.elts)
+        if isinstance(ast_node, ast.Dict):
+            return {
+                cls._convert_ast_to_py(k): cls._convert_ast_to_py(v)
+                for k, v in zip(ast_node.keys, ast_node.values)
+            }
+        raise ValueError(f"Unsupported type {type(ast_node)}")
+
+
+class ConvertClass(AstDecorator):
+    """
+    Class decorator for guiding generation of sync classes
+
+    Args:
+        sync_name: use a new name for the sync class
+        replace_symbols: a dict of symbols and replacements to use when generating sync class
+        docstring_format_vars: a dict of variables to replace in the docstring
+        rm_aio: if True, automatically strip all asyncio keywords from method. If false,
+            only keywords wrapped in CrossSync.rm_aio() calls to be removed.
+        add_mapping_for_name: when given, will add a new attribute to CrossSync,
+            so the original class and its sync version can be accessed from CrossSync.<name>
+    """
+
+    def __init__(
+        self,
+        sync_name: str | None = None,
+        *,
+        replace_symbols: dict[str, str] | None = None,
+        docstring_format_vars: dict[str, tuple[str | None, str | None]] | None = None,
+        rm_aio: bool = False,
+        add_mapping_for_name: str | None = None,
+    ):
+        self.sync_name = sync_name
+        self.replace_symbols = replace_symbols
+        docstring_format_vars = docstring_format_vars or {}
+        self.async_docstring_format_vars = {
+            k: v[0] or "" for k, v in docstring_format_vars.items()
+        }
+        self.sync_docstring_format_vars = {
+            k: v[1] or "" for k, v in docstring_format_vars.items()
+        }
+        self.rm_aio = rm_aio
+        self.add_mapping_for_name = add_mapping_for_name
+
+    def async_decorator(self):
+        """
+        Use async decorator as a hook to update CrossSync mappings
+        """
+        from .cross_sync import CrossSync
+
+        if not self.add_mapping_for_name and not self.async_docstring_format_vars:
+            # return None if no changes needed
+            return None
+
+        new_mapping = self.add_mapping_for_name
+
+        def decorator(cls):
+            if new_mapping:
+                CrossSync.add_mapping(new_mapping, cls)
+            if self.async_docstring_format_vars:
+                cls.__doc__ = cls.__doc__.format(**self.async_docstring_format_vars)
```

Comment:
what does this do? why is this only formatting async_docstring? How about sync_docstring?

Reply:
CrossSync allows you to ad variables into docstrings, and then it will replace them with a different string for sync vs async. This came up because some async methods raise exceptions when run outside of an event loop context, but that doesn't apply to sync versions. So the docstrings are mostly the same, but with an extra little note for async

On the sync side, we can replace the variables [when generating the sync file](https://github.com/googleapis/python-bigtable/blob/480b139444f69cf143523331e5c1ae0982674f71/google/cloud/bigtable/data/_cross_sync/_decorators.py#L284). On the async side, we update the `__doc__` attribute of the class as part of the class decoration process. [The code itself will have template variables](https://github.com/googleapis/python-bigtable/blob/2ab2e40827fe5f9a0231db7c266f81a39e30fdce/google/cloud/bigtable/data/_async/client.py#L143) when you read it directly, but docs generators will read from `class.__doc__` when generating content